ibis/expr/analysis.py

```diff
@@ -1058,29 +1036,13 @@ def find_source_table(expr):
     ...
     NotImplementedError: More than one base table not implemented
     """
-    first_tables = []
-
-    stack = [expr]
-    seen = set()
-
-    while stack:
-        e = stack.pop()
-        op = e.op()
-
-        if op not in seen:
-            seen.add(op)
-
-            arguments = [
-                arg for arg in reversed(list(op.flat_args()))
-                if isinstance(arg, ir.Expr)
-            ]
-            first_tables.extend(
-                arg for arg in arguments if isinstance(arg, ir.TableExpr)
-            )
-            stack.extend(
-                arg for arg in arguments if not isinstance(arg, ir.TableExpr)
-            )
+    def finder(expr):
+        if isinstance(expr, ir.TableExpr):
+            return lin.stop, expr
+        else:
+            return lin.proceed, None
```

Comment:
So this would look like the following?

```python
lin.traverse(
    lambda expr: expr if isinstance(expr, ir.TableExpr) else None, 
    expr.op().flat_args()
    halt=lambda expr, result: isinstance(expr, ir.TableExpr)
)
```

Reply:
Hm, now I'm not so sure I like my suggestion. This duplicates the condition which isn't good.

Reply:
Let's keep your API for now.

Reply:
Ok, should I rename `stop` to `halt`?

Reply:
Yep, then will merge